### PR TITLE
Make isinstance check pass for proxy ndarrays

### LINKED ISSUE #14537: [FEA] Proxy ndarrays don't pass an `instancecheck()` for `np.ndarray`
There's a lot of code out there that does something like:

```python
if not isinstance(x, np.ndarray):
    raise TypeError("Not a numpy array")
```

This is a problem for `cudf.pandas`, because proxy ndarray types, such as those returned by `pd.Series.values` do not pass the `isinstance()` check.

Ideally, more projects would avoid doing a hard `isinstance()` check and instead use something like EAFP:

```python
np.asarray(x)
```

..but that is not the world we live in today. Too many third-party libraries that people want to use with `cudf.pandas` use the pattern above, so it's on us to solve the problem right now.

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Closes #14537.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/pandas/proxy_base.py
Comment (Matt711) [inline python/cudf/cudf/pandas/proxy_base.py:10]: I'm returning an empty ndarray which is causing numba dispatched functions to return incorrect results ie.

```python
@njit(nogil=True)
def f(x):
    return 2*x

x = pd.Series([1]).values

print(f(x))
```

I need to somehow make sure that `_fsproxy_wrapped` attribute (which contains the real numpy or cupy array) is used. Or go with a different approach.

**Comment by Matt711** [comment]
This PR fixes the two issues with #16286 (ie. the previous version of this PR that we reverted). The two issues were:

1. In the `__array_finalize__(self, obj)`, the `_fsproxy_wrapped` attribute  is set equal to `obj`, which could be `None`. In this PR, there's a check `if (obj is None): return` to prevent that. I ran the cudf-pandas-integration tests, and it seems to fix the tests that were failing from the previous version of this PR. Once #16645 is merged , we'll be able to see the third-party integration tests passing in cudf.

2. In `__new__`, there was an initial DtoH transfer when a `cp.ndarray` was passed to the proxy array class. This PR doesn't include the DtoH transfer. Instead, in `__new__`, we're creating a "garbage" `np.ndarray` with the correct shape and dtype, but with incorrect/garbage data.

The fix for 2. leads to a problem which the only way I can think of solving is via monkeypatching. The problem is that now that our proxy array passes the check `isinstance(proxy, np.ndarray)`, there are many numpy functions (which _will_ call NumPy C functions) that will assume that they can use the array's buffer instead instead of going through a double underscore method like `__array__`. This is a problem because our proxy array's buffer has garbage data, so the function will produce garbage data.

Take `np.asarray` for an example. It will eventually lead to some like code like this. Our proxy array passes the `PyArray_Check` but gives incorrect data in the `PyArray_DATA` call.
```c
    if (PyArray_Check(obj)) {
        void* buffer = PyArray_DATA(obj) // access the buffer directly
    }
```

I think the only way to avoid this problem from Python is to monkeypatch every function like `np.asarray`. So far I've found several functions we'll need to patch: `np.asarray`, `np.where`, `np.outer`, `np.inner`, and several other functions in the `np.linalg` module.

There is at least one other function in a third-party library which we'll need to patch (`torch.from_numpy`). This function function could be patched in the same way the numpy functions are.

Numba CPU dispatched functions also access the proxy arrays buffer. The good news is that eventually numba will support compiling objects which implement `__array__` (see [numba/9584](https://github.com/numba/numba/pull/9584)). This bad news is that this probably won't be ready before 24.10. So in the meantime, we would have to let the user know that using proxy arrays directly in CPU dispatched functions will return incorrect results. If there's not a way to fail loudly in this case, then it should at least be documented. FYI, we currently fail when using proxy arrays in CPU dispatched functions because we fail isinstance check. So if its possible to continue failing, that's not the worst thing to until the next Numba release.

What do you all think of the monkey patching approach? 

Another idea would be pay the cost of the DtoH transfer upfront (like in the previous PR). The DtoH is only done on instance creation. And then after that, the fast-slow proxy mechanism is used. The pro here is that the buffer will be set correctly, and thus no monkeypatching. The con is obviously the DtoH transfer.

**Comment by vyasr** [comment]
Summary of offline discussion: monkey-patching numpy to make this work is a bridge too far without much stronger motivations, and we will probably just go with the eager D2H copy instead.

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py
Comment (Matt711) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1700]: TODO: Move the tests once #16645 is merged

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (Matt711) [inline python/cudf/cudf/pandas/_wrappers/numpy.py:125]: For custom ufuncs, result (which is a proxy array) has a dtype of object. I'm running `np.asarray` over the result to get the right dtype. I think the dtype is getting lost somewhere in one of the calls to `_fast_slow_function_call` (likely after falling back).

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (galipremsagar) [inline python/cudf/cudf/pandas/_wrappers/numpy.py:117]: Are there ufuncs that return a tuple of ndarrays?

Comment (Matt711) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:117]: Yes `np.modf`and `np.frexp` for sure. But there are probably others as well

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (mroeschke) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:125]: Could you give an example where this happens? Are you suggesting this could be happening when `cupy.ndarray.get` is called to convert the cupy array to a numpy array?

Comment (Matt711) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:125]: I can continue looking into this and see what changes (if any) can be made in to [`fast_slow_proxy.py`](https://github.com/rapidsai/cudf/blob/branch-24.10/python/cudf/cudf/pandas/fast_slow_proxy.py) to ensure the dtype remains after fallback (eg. in this case it should be `int64`). And make them in a follow-up PR

Comment (mroeschke) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:125]: > Are you suggesting this could be happening when cupy.ndarray.get is called to convert the cupy array to a numpy array?

Yeah it was just a guess. I'm just surprised that

1. There is an intermediate state where this array is `object`
2. `np.asarray` infers an `object` type to a `non-object` type e.g.
```python
In [1]: import numpy as np

In [2]: arr = np.array([1, 2], dtype=object)

In [3]: np.asarray(arr)
Out[3]: array([1, 2], dtype=object)
```

Comment (Matt711) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:125]: Thanks for looking at this too.

> 1. There is an intermediate state where this array is `object`

I added some print calls to `_fast_slow_function_call` to understand better what's happening. On one call to `_fast_slow_function_call` I see
```
FAST ARGS (<class 'cupy.ndarray'>, 'flags')
```
And on the next call I see. I guess there's something going on with `flags`, but I'm unsure of the root cause yet.
```
FALLBACK Unsupported dtype object
SLOW ARGS (array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=object), 'flags')
```
> 2. `np.asarray` infers an `object` type to a `non-object` type e.g.

I think this is because in `__new__` were creating a proxy array instance with the right dtype. Eg.
```
obj = np.asarray(arr.get(), dtype=arr.dtype).view(cls)
```
And `np.asarray` returns this value, not the array in `_fsproxy_wrapped` which would still have dtype==`object` I think.

**Review by galipremsagar — APPROVED** [review]
Apart from this issue: https://github.com/rapidsai/cudf/pull/16601/files#r1736269538 for which @Matt711 can investigate and issue a follow-up fix, I'm happy with the current state of changes.

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/_wrappers/numpy.py:121]:
```suggestion
            return tuple(numpy.asarray(x) for x in result)
```

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (mroeschke) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:125]: Ok thanks for looking a bit deeper. We do have a proxy object for flags `_ndarray_flags` below. Do you think it's hitting that proxy object?

> I think this is because in __new__ were creating a proxy array instance with the right dtype. Eg.

Ah gotcha. If you're not able to identify the root cause immediately, we can punt to a follow up PR, but it would be good to avoid this object casting because this is making an unnecessary copy (given the original data was not object type)

**Review by mroeschke — APPROVED** [review]

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (Matt711) [reply python/cudf/cudf/pandas/_wrappers/numpy.py:125]: > Ok thanks for looking a bit deeper. We do have a proxy object for flags `_ndarray_flags` below. Do you think it's hitting that proxy object?

I think it is. Right now I'm using print statements to debug the problem, which is tedious. I'm doing this because I'm having trouble setting breakpoints in `_fast_slow_function_call` that are caught by gdb. I'll ask in the team chat about this next week.
 
> > I think this is because in **new** were creating a proxy array instance with the right dtype. Eg.
> 
> Ah gotcha. If you're not able to identify the root cause immediately, we can punt to a follow up PR, but it would be good to avoid this object casting because this is making an unnecessary copy (given the original data was not object type)

I think the default for `np.asarray` is `copy=None` (not `copy=True`) which means [a copy is only performed as needed](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html#numpy.asarray). I thought we might be able to get away with `copy=False` but there are cases like [generalized ufuncs](https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html) which have to perform a copy and `copy=False` fails loudly. So I think this might work in the intermediate term because a copy usually won't happen.

**Comment by Matt711** [comment]
Ping @vyasr for a review next week

**Comment by bdice** [comment]
@Matt711 The PR description says "do not merge" but there is no "DO NOT MERGE" label. Can you make this consistent?

Also for team knowledge, the "Description" section of the PR body is used in the final commit message when the PR is merged. Temporary information like the PR state or benchmarks are better to put in comments rather than the "Description" section.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (vyasr) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:572]: Can we explicitly match the type rather than a blanket else statement, or write a helper singledispatch function to accomplish the same (not sure if singledispatch will work with proxies though)? As soon as a new type is added this will fail in a potentially confusing way depending on how the arguments are handled.

File: python/cudf/cudf/pandas/proxy_base.py
Comment (vyasr) [inline python/cudf/cudf/pandas/proxy_base.py:20]:
```suggestion
        if isinstance(arr, cp.ndarray):
            arr = arr.get()
        if not isinstance(arr, np.ndarray):
            raise TypeError(
                "Unsupported array type. Must be numpy.ndarray or cupy.ndarray"
            )
        return np.asarray(arr, dtype=arr.dtype).view(cls)
```

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py
Comment (vyasr) [reply python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1736]: Presumably you were planning on doing this in a follow-up (fine with me).

**Review by vyasr — APPROVED** [review]
Couple of suggestions for improvement, but assuming you don't object to applying them I don't need to review again.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:572]: The singledispatch idea is interesting, I'll give it a try. But will likely go with your first suggestion

File: .github/workflows/pr.yaml
Comment (Matt711) [inline .github/workflows/pr.yaml:308]: Don't Review. I'm going to remove once I confirm job passes

**Pull request merged**
